bundle/file.go

```diff
@@ -265,3 +269,93 @@ func (t *tarballLoader) NextFile() (*Descriptor, error) {
 
 	return newDescriptor(path.Join(t.baseURL, f.name), f.name, f.reader), nil
 }
+
+// Next implements the storage.Iterator interface.
+// It iterates to the next policy or data file in the directory tree
+// and returns a storage.Update for the file.
+func (t *tarballLoader) Next() (*storage.Update, error) {
+	if t.tr == nil {
+		gr, err := gzip.NewReader(t.r)
+		if err != nil {
+			return nil, errors.Wrap(err, "archive read failed")
+		}
+
+		t.tr = tar.NewReader(gr)
+	}
+
+	if t.files == nil {
+		t.files = []file{}
+
+		for {
+			header, err := t.tr.Next()
+			if err == io.EOF {
+				break
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			// Keep iterating on the archive until we find a normal file
+			if header.Typeflag == tar.TypeReg {
+
+				if filepath.Base(header.Name) == dataFile || filepath.Base(header.Name) == yamlDataFile || strings.HasSuffix(header.Name, RegoExt) {
+					f := file{name: header.Name}
+
+					var buf bytes.Buffer
+					if _, err := io.Copy(&buf, t.tr); err != nil {
+						return nil, errors.Wrapf(err, "failed to copy file %s", header.Name)
+					}
+
+					f.reader = &buf
+
+					t.files = append(t.files, f)
+				}
+			}
+		}
+		sortFilePathAscend(t.files)
+	}
+
+	// If done reading files then just return io.EOF
+	// errors for each NextFile() call
+	if t.idx >= len(t.files) {
+		return nil, io.EOF
+	}
+
+	f := t.files[t.idx]
+	t.idx++
+
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, f.reader); err != nil {
+		return nil, errors.Wrapf(err, "failed to copy file %s", f.name)
+	}
+
+	isPolicy := false
+	fpath := strings.TrimLeft(filepath.ToSlash(filepath.Dir(f.name)), "/.")
+
+	if strings.HasSuffix(f.name, RegoExt) {
+		isPolicy = true
+		fpath = strings.Trim(f.name, "/")
+	}
+
+	return &storage.Update{
+		Path:     fpath,
+		Value:    buf.Bytes(),
+		IsPolicy: isPolicy,
+	}, nil
+}
+
+func NewIterator(r io.Reader) storage.Iterator {
+	it := tarballLoader{
+		r: r,
+	}
+	return &it
+}
+
+func sortFilePathAscend(files []file) {
+	sort.Slice(files, func(i, j int) bool {
+		a := strings.TrimLeft(filepath.ToSlash(filepath.Dir(files[i].name)), "/.")
```

Comment:
Let's sort based on the storage paths that are specified in the write calls so that if there's a difference between those paths and the raw file paths, the sorting takes that into account.

Reply:
Updated.

Reply:
> I was somewhat surprised to see the use of Tee here. I would have expected us to the read the files into memory and then hang onto []byte for the data files that need to be lazily loaded. The current implementation isn't asymptotically worse in terms of memory/cpu usage, however, it does mean that the bundle serialization format (e.g., tar/gz) is exposed to the caller--they're not just dealing with a logical path => []byte map anymore. Perhaps a comment explaining why this approach was selected would help.

@tsandall the code has been updated to hold the policy/data bytes instead of the serialized tar.gz.